### PR TITLE
Stricter C99 warnings, generate primitive list with types

### PR DESCRIPTION
**Pull request by MisterDA**
I was working on this at the end of last week after reading [Modernizing Fedora's C code](https://lwn.net/Articles/913505/), and over the week-end #11759 was opened which this PR should fix.

This PR:
- enables new warning for stricter C99 generation;
- fixes strict-prototypes and old-style-definition warnings;
- generates the list of primitives with their argument list.

It generates the list of primitives with their arguments by running first the C preprocessor on C files containing primitives, extracting the functions with a sed script. It retains the caching behaviour introduced in #8696. It also introduces `PrimitiveX` macros for all primitive arities.

Possible follow-ups for this PR include checking support from the compiler for the warnings, defining a better `CAMLprim` macro _declaring_ the arity of the function and possibly using it at runtime to check the arity of a called function.

### CODE REVIEW

**Comment by MisterDA** [comment]
@xavierleroy was faster than me with #11762 which is equivalent to my 54210b5b.

**Comment by xavierleroy** [comment]
Could you please submit 54210b5bf3 (the middle commit) as a separate PR, so that it can be merged immediately?  (Plus, it subsumes my PR #11762, so I'll be able to close #11762.)

**Comment by xavierleroy** [comment]
I think your 54210b5 is more comprehensive than my #11762, so let's use your code instead.

**Comment by MisterDA** [comment]
Submitted this commit as #11764. I've used your commit message instead of mine.

**Comment by dra27** [comment]
At the moment, this goes to considerable lengths to generate accurate prototypes for the primitives. Those prototypes are then forcibly cast to `c_primitive` in order to put them all in `caml_builtin_cprim`. Wouldn't it be easier just to declare them all as `caml_prim(void)` in `prims.c`? There's a tiny gotcha because `runtime/caml/mlvalues.h` contains the actual prototypes for `caml_set_oo_name` and `caml_get_public_method` but this can be relatively easier worked around to get this patch which when combined with just `runtime/interp.c` and `runtime/caml/prims.h` from 63cc8272c7e2723b801c98749accaed1cdf7a279 also builds without triggering any warnings:

```diff
--- a/Makefile
+++ b/Makefile
@@ -771,10 +771,10 @@ runtime/primitives: \
        cp $^ $@

 runtime/prims.c : runtime/primitives
-       (echo '#define CAML_INTERNALS'; \
-         echo '#include "caml/mlvalues.h"'; \
-        echo '#include "caml/prims.h"'; \
-        sed -e 's/.*/extern value &();/' $<; \
+       (echo '#include "caml/config.h"'; \
+        echo 'typedef intnat value;'; \
+        echo 'typedef value (*c_primitive)(void);'; \
+        sed -e 's/.*/extern value &(void);/' $<; \
         echo 'c_primitive caml_builtin_cprim[] = {'; \
         sed -e 's/.*/  &,/' $<; \
         echo '  0 };'; \
```

File: Makefile [hunk 1]
```diff
@@ -762,24 +769,43 @@ runtime/ld.conf: $(ROOTDIR)/Makefile.config
...
+# To speed up builds, we avoid changing "primitives" and
+# "primitive-arguments" when files containing primitives change but
+# the primitives table does not
+runtime/primitives-arguments.new: \
+  $(runtime_BUILT_HEADERS) \
+  $(runtime_PRIMITIVES) \
+  runtime/gen_primitives.sed runtime/gen_primitives_ints.sed
+	(for prim in $(runtime_PRIMITIVES); do \
+	     $(CPP) -I./runtime -D'CAMLprim()=0' "$${prim}" | \
+	     sed -n -f runtime/gen_primitives.sed; \
```
Comment (xavierleroy) [inline Makefile:779]: Is the preprocessing really needed?

File: runtime/caml/prims.h [hunk 2]
```diff
@@ -30,7 +30,15 @@ extern struct ext_table caml_prim_table;
...
-#define Primitive(n) ((c_primitive)(caml_prim_table.contents[n]))
+#define Primitive1(n) ((value (*)(value))caml_prim_table.contents[n])
+#define Primitive2(n) ((value (*)(value,value))caml_prim_table.contents[n])
+#define Primitive3(n) \
+    ((value (*)(value,value,value))caml_prim_table.contents[n])
+#define Primitive4(n) \
+    ((value (*)(value,value,value,value))caml_prim_table.contents[n])
+#define Primitive5(n) \
+    ((value (*)(value,value,value,value,value))caml_prim_table.contents[n])
+#define PrimitiveN(n) ((value (*)(value*,int))caml_prim_table.contents[n])
```
Comment (xavierleroy) [inline runtime/caml/prims.h]: Do yourself a favor and introduce typedefs:
```
typedef value (*c_primitive1)(value);
...
#define Primitive1(n) (c_primitive1)(caml_prim_table.contents[n]))
```

File: runtime/caml/prims.h [hunk 3]
```diff
@@ -20,7 +20,7 @@
 
 #ifdef CAML_INTERNALS
 
-typedef value (*c_primitive)();
+typedef value (*c_primitive)(void);
```
Comment (xavierleroy) [inline runtime/caml/prims.h]: You could just as well use
```
typedef void * c_primitive;
```
The built-in primitives from table `caml_builtin_cprim` are never called directly anyway, just added to an extensible table with type `void *`.

(Note that ISO C forbids casting a function pointer to type `void *`, but we've been doing this for a very long time, and all C compilers that I know of are happy with this.)

File: runtime/gen_primitives.sed [hunk 4]
```diff
@@ -15,22 +13,13 @@
...
 #**************************************************************************
 
-# duplicated from $(ROOTDIR)/runtime/Makefile
+/^CAMLprim value/ {
+  :loop
+  /)/ ! {
+    N; b loop
+  }
+
+  s/\n  */ /g
```
Comment (xavierleroy) [inline runtime/gen_primitives.sed:22]: I didn't know you had conditional jumps in sed...  This seems to be in POSIX sed, even.  

This said, if you're willing to suppose primitive definitions take at most 2 lines, there's a loop-free approach:
on seeing `/^CAMLprim /`, fetch next line with `N`, then remove everything following the first `)`.

File: runtime/sys.c [hunk 5]
```diff
@@ -219,7 +219,7 @@ CAMLprim value caml_sys_exit(value retcode)
 #endif
 #endif
 
-const static int sys_open_flags[] = {
+static const int sys_open_flags[] = {
```
Comment (xavierleroy) [inline runtime/sys.c:222]: Unrelated change.

File: Makefile [hunk 6]
```diff
@@ -762,24 +769,43 @@ runtime/ld.conf: $(ROOTDIR)/Makefile.config
...
 	 echo '#include "caml/prims.h"'; \
-	 sed -e 's/.*/extern value &();/' $<; \
+	 echo 'struct lexer_buffer; struct lexing_table;'; \
+	 echo 'struct parser_env; struct parser_tables;'; \
+	 echo; \
+	 sed -e 's/.*/extern value &;/' $<; \
+	 echo; \
 	 echo 'c_primitive caml_builtin_cprim[] = {'; \
-	 sed -e 's/.*/  &,/' $<; \
+	 sed -e 's/^\([a-z0-9_][a-z0-9_]*\) *(.*)/  (c_primitive)\1,/' $<; \
```
Comment (xavierleroy) [inline Makefile]: With my suggestion `typedef void * c_primitive;`, the cast to `(c_primitive)` is not needed.

File: Makefile [hunk 7]
```diff
@@ -762,24 +769,43 @@ runtime/ld.conf: $(ROOTDIR)/Makefile.config
...
 	 echo '  0 };'; \
+	 echo; \
 	 echo 'char * caml_names_of_builtin_cprim[] = {'; \
-	 sed -e 's/.*/  "&",/' $<; \
+	 sed -e 's/^\([a-z0-9_][a-z0-9_]*\) *(.*)/  "\1",/' $<; \
```
Comment (xavierleroy) [inline Makefile:808]: All this `sed` without setting `LC_CTYPE` makes me nervous.  Experience shows that it fails with bizarre locales.  Or, use named character classes, e.g. `[[:alnum]_]`.

File: Makefile [hunk 8]
```diff
@@ -762,24 +769,43 @@ runtime/ld.conf: $(ROOTDIR)/Makefile.config
...
-  $(shell runtime/gen_primitives.sh > runtime/primitives.new; \
-                    cmp -s runtime/primitives runtime/primitives.new || \
-                    echo runtime/primitives.new)
-	cp $^ $@
+# To speed up builds, we avoid changing "primitives" and
+# "primitive-arguments" when files containing primitives change but
+# the primitives table does not
+runtime/primitives-arguments.new: \
+  $(runtime_BUILT_HEADERS) \
+  $(runtime_PRIMITIVES) \
```
Comment (xavierleroy) [inline Makefile:774]: You're moving code from a script into the toplevel Makefile.  I'm not sure this is a good idea.

**Review by xavierleroy — COMMENTED** [review]
Some suggestions and remarks.  I admire your mastery of `sed`, but would still prefer a more civilized scripting language.

File: Makefile (hunk 1 above)
Comment (MisterDA) [reply Makefile:779]: The preprocessor is needed for some symbols in `runtime/floats.c`:

https://github.com/ocaml/ocaml/blob/d0cdf4b0a075ddb56a5523c4d51c99408321f609/runtime/floats.c#L1127-L1131

I couldn't extract the arguments without expanding the macros. Using the preprocessor also removes this concern (I forgot to remove the comment):

```
# If primitives contain duplicated lines (e.g. because the code is defined
# like
# #ifdef X
# CAMLprim value caml_foo() ...
# #else
# CAMLprim value caml_foo() ...
# #endif)
```

File: runtime/sys.c (hunk 5 above)
Comment (MisterDA) [reply runtime/sys.c:222]: This fixes the `-Wold-style-declaration` warning that I couldn't enable globally, as it's specific to GCC.

File: Makefile (hunk 8 above)
Comment (MisterDA) [reply Makefile:774]: It should be ok: we switch from one trick to use a script to conditionally add a prerequisite or not, which felt like advanced magic, to resetting the modified time of a file.
It does break the Dune build, though.

File: runtime/caml/prims.h (hunk 2 above)
Comment (MisterDA) [reply runtime/caml/prims.h]: That's much more readable, thanks.

File: runtime/caml/prims.h (hunk 3 above)
Comment (MisterDA) [reply runtime/caml/prims.h]: Also nice, thanks.

File: Makefile (hunk 6 above)
Comment (MisterDA) [reply Makefile]: Applied, thanks.

File: Makefile (hunk 7 above)
Comment (MisterDA) [reply Makefile:808]: The comment above suggested `LC_ALL=C` instead of `LC_CTYPE`. I'm not familiar with these, which one should I use? In the meantime, I've using `export LC_ALL=C;` where needed.

File: Makefile (hunk 1 above)
Comment (xavierleroy) [reply Makefile:779]: Good points, thanks for the explanations.  The `DEFINE_NAN_CMP` macro in `runtime/floats.c` is ugly as hell and could be rewritten.  For the alternate definitions, we used to rely on `sort | uniq` to keep only one, but it gets more fragile when the prototypes are taken into account, that much is true.

File: Makefile (hunk 8 above)
Comment (xavierleroy) [reply Makefile:774]: The Dune build could be a concern.  But my more general concern is that too much scripting within a Makefile renders the Makefile unreadable; at some point, having separate scripts that do one thing but do it well becomes preferable.

Comment (MisterDA) [reply Makefile:774]: I agree, that's why I've externalised the sed scripts.
However if we chose to keep using the preprocessor, I think it'll be difficult to give the `$(CPP)` content to the script due to usual concerns of spaces and word-splitting.

**Comment by xavierleroy** [comment]
Concerning @dra27's suggestion:

> At the moment, this goes to considerable lengths to generate accurate prototypes for the primitives. Those prototypes are then forcibly cast to c_primitive in order to put them all in caml_builtin_cprim. Wouldn't it be easier just to declare them all as caml_prim(void) in prims.c? There's a tiny gotcha because runtime/caml/mlvalues.h contains the actual prototypes for caml_set_oo_name and caml_get_public_method but this can be relatively easier worked around […]

On the one hand, it's not legal C11 to declare `value f(void)` a function `f` that is defined in another compilation unit as `value f(value)`, say; while I think it is correct to declare it `value f()`, like we currently do.  On the other hand, this is no worse than the many other "not legal C11" things we already do,  and it does work around the problem with great economy of means.  So maybe we should commit this solution (plus the modified prims.h and interp.c files) first.

File: runtime/gen_primitives.sed (hunk 4 above)
Comment (MisterDA) [reply runtime/gen_primitives.sed:22]: I could, but `caml_array_fill` would disagree!
https://github.com/ocaml/ocaml/blob/ac52c3ab4a545d77eea034993a0947ba5e79762b/runtime/array.c#L557-L560
I think there might be fewer surprises (apart from the script itself) if we loop in sed.

**Comment by MisterDA** [comment]
Thank you both for these rounds of comprehensive reviews. As it seems that keeping the primitives prototypes is not useful right now (which is on me, I may have dove to deep in sed and Makefile scripting), I'm closing this PR. I'll submit a simpler one soon, without the full prototypes and the preprocessor pass. We can always revisit this idea later.
